drivers/watchdog/wdt_andes_atcwdt200.c

```diff
@@ -0,0 +1,371 @@
+
+	reg = sys_read32(WDT_CTRL(dev));
+	reg &= ~(WDT_CTRL_RSTEN | WDT_CTRL_EN);
+
+	sys_write32(WDT_WREN_NUM, WDT_WREN(dev));
+	sys_write32(reg, WDT_CTRL(dev));
+
+	k_spin_unlock(&data->lock, key);
+
+	wdt_atcwdt200_set_max_timeout(dev);
+	counter_cancel_channel_alarm(counter_dev, 2);
+
+	return 0;
+}
+
+static int wdt_atcwdt200_setup(const struct device *dev, uint8_t options)
+{
+	struct wdt_atcwdt200_dev_data *data = dev->data;
+	k_spinlock_key_t key;
+	uint32_t reg;
+	uint32_t ret = 0;
+
+	if (!data->timeout_valid) {
+		LOG_ERR("No valid timeouts installed");
+		return -EINVAL;
+	}
+
+	key = k_spin_lock(&data->lock);
+
+	reg = sys_read32(WDT_CTRL(dev));
+	reg |= (WDT_CTRL_RSTEN | WDT_CTRL_EN);
+
+	if ((options & WDT_OPT_PAUSE_HALTED_BY_DBG) ==
+			WDT_OPT_PAUSE_HALTED_BY_DBG) {
+		counter_cancel_channel_alarm(counter_dev, 2);
+		sys_write32(WDT_WREN_NUM, WDT_WREN(dev));
+		sys_write32(reg, WDT_CTRL(dev));
+		goto out;
+	} else {
+		ret = counter_set_channel_alarm(counter_dev, 2, &alarm_cfg);
+		if (ret != 0) {
+			ret = -EINVAL;
+			goto out;
+		}
+
+		sys_write32(WDT_WREN_NUM, WDT_WREN(dev));
+		sys_write32(reg, WDT_CTRL(dev));
+	}
+
+out:
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+/**
+ * @brief Calculates the watchdog counter value (wdogcmp0) and
+ *        scaler (wdogscale) to be installed in the watchdog timer
+ *
+ * @param timeout Timeout value in milliseconds.
+ * @param clk     Clock of watchdog in Hz.
+ * @param scaler  Pointer to return scaler power of 2
+ *
+ * @return Watchdog counter value
+ */
+static uint32_t wdt_atcwdt200_convtime(uint32_t timeout, uint32_t *scaler)
+{
+	int i;
+	uint32_t rst_period, cnt;
+
+	cnt = (uint32_t)((timeout * EXT_CLOCK_FREQ) / 1000);
+	rst_period = cnt;
+
+	for (i = 0; i < 14 && cnt > 0; i++) {
+		cnt >>= 1;
+	}
+
+	*scaler = i;
+
+	return rst_period;
+}
+
+static int wdt_atcwdt200_install_timeout(const struct device *dev,
+				      const struct wdt_timeout_cfg *cfg)
+{
+	struct wdt_atcwdt200_dev_data *data = dev->data;
+	k_spinlock_key_t key;
+	uint32_t rst_period, reg, counter_freq, scaler;
+
+	if (cfg->window.min != 0U || cfg->window.max == 0U) {
+		return -EINVAL;
+	}
+
+	counter_freq = counter_get_frequency(counter_dev);
+	rst_period = wdt_atcwdt200_convtime(cfg->window.max, &scaler);
+
+	if (rst_period < 0 || WDOGCFG_PERIOD_MAX < rst_period) {
+		LOG_ERR("Unsupported watchdog timeout\n");
+		return -EINVAL;
+	}
+
+	wdt_atcwdt200_disable(dev);
+
+	key = k_spin_lock(&data->lock);
+
+	switch (cfg->flags) {
+	case WDT_FLAG_RESET_SOC:
+		if (scaler < 7) {
+			reg = WDT_CTRL_RSTTIME_POW_2_7;
+		} else {
+			scaler = scaler - 7;
+			reg = scaler << 8;
+		}
+
+		alarm_cfg.flags = 0;
+		alarm_cfg.callback = wdt_counter_cb;
+		alarm_cfg.user_data = &alarm_cfg;
+		alarm_cfg.ticks = (((cfg->window.max * counter_freq) / 1000) >> 1);
+
+		break;
+	case WDT_FLAG_RESET_NONE:
+	case WDT_FLAG_RESET_CPU_CORE:
+	default:
+		LOG_ERR("Unsupported watchdog config flags\n");
+		return -ENOTSUP;
```

Comment:
`k_spin_unlock` missing here

Reply:
Oh, sorry for that.
I have update the source code

Thanks.